File: sdk-java/src/main/java/io/superdurable/dex/ClientOptions.java
/*
 * Legacy Materials in this file remain under their original licenses.
 * See LEGACY_NOTICES.md.
 */

/*
 * Modifications Copyright (c) 2026 Super Durable, Inc.
 *
 * Modifications after the Legacy Cutoff are licensed under the
 * Super Durable Source License 1.0.
 * Legacy Materials remain under their original licenses.
 * See LICENSE and LEGACY_NOTICES.md.
 */

package io.superdurable.dex;

import com.fasterxml.jackson.databind.ObjectMapper;

/**
 * Configures a synchronous {@link Client} connection and value mapper.
 *
 * <p>The client connects to the Dex server over plaintext gRPC. Its default server address is
 * {@code localhost:8801}. A worker target is copied into start-time Flow configuration unless a
 * request supplies a more specific target. The default Jackson mapper discovers installed modules;
 * pass a configured mapper when application values require custom serialization.
 *
 * <pre>{@code
 * ClientOptions options = new ClientOptions(
 *         "dex.example.com:8801",
 *         new WorkerTarget("orders-worker:8803", false),
 *         applicationObjectMapper);
 * }</pre>
 */
public final class ClientOptions {
    private final String serverAddress;
    private final WorkerTarget workerTarget;
    private final ObjectMapper objectMapper;

    /** Creates options for {@code localhost:8801} with the default Jackson mapper. */
    public ClientOptions() {
        this("localhost:8801", null, new ObjectMapper().findAndRegisterModules());
    }

    /**
     * Creates options for a Dex server with default worker routing and serialization.
     *
     * @param serverAddress the gRPC target understood by {@code ManagedChannelBuilder}
     */
    public ClientOptions(final String serverAddress) {
        this(serverAddress, null, new ObjectMapper().findAndRegisterModules());
    }

    /**
     * Creates options with explicit server and default worker routing.
     *
     * @param serverAddress the Dex gRPC target
     * @param workerTarget the default Flow worker target, or {@code null}
     */
    public ClientOptions(
            final String serverAddress,
            final WorkerTarget workerTarget) {
        this(serverAddress, workerTarget, new ObjectMapper().findAndRegisterModules());
    }

    /**
     * Creates options with explicit transport, routing, and serialization settings.
     *
     * @param serverAddress the Dex gRPC target
     * @param workerTarget the default Flow worker target, or {@code null}
     * @param objectMapper the nonnull Jackson mapper used for application values
     * @throws IllegalArgumentException if {@code objectMapper} is {@code null}
     */
    public ClientOptions(
            final String serverAddress,
            final WorkerTarget workerTarget,
            final ObjectMapper objectMapper) {
        if (objectMapper == null) {
            throw new IllegalArgumentException("objectMapper is required");
        }
        this.serverAddress = serverAddress;
        this.workerTarget = workerTarget;
        this.objectMapper = objectMapper;
    }

    String getServerAddress() {
        return serverAddress;
    }

    WorkerTarget getWorkerTarget() {
        return workerTarget;
    }

    ObjectMapper getObjectMapper() {
        return objectMapper;
    }
}

File: sdk-rust/crates/dex-sdk/src/client_options.rs
// Copyright (c) 2026 Super Durable, Inc.
//
// Licensed under the Super Durable Source License 1.0.
// You may not use this file except in compliance with the License.
// See the LICENSE file in the repository root.
//
// SPDX-License-Identifier: LicenseRef-Super-Durable-1.0

use crate::WorkerTarget;

#[derive(Clone, Debug)]
/// Configures how a [`crate::Client`] reaches Dex and identifies application workers.
///
/// [`ClientOptions::new`] connects to `127.0.0.1:8801` over plaintext gRPC and supplies no default
/// worker target. A per-Flow [`crate::FlowConfig`] can override the target.
pub struct ClientOptions {
    server_address: String,
    worker_target: Option<WorkerTarget>,
}

impl ClientOptions {
    /// Creates options with the local Dex server default and no worker target.
    pub fn new() -> Self {
        Self {
            server_address: "127.0.0.1:8801".to_string(),
            worker_target: None,
        }
    }

    /// Sets the plaintext gRPC `host:port` used for every client request.
    pub fn server_address(mut self, value: impl Into<String>) -> Self {
        self.server_address = value.into();
        self
    }

    /// Sets the WorkerService target advertised by newly started Flows.
    pub fn worker_target(mut self, value: WorkerTarget) -> Self {
        self.worker_target = Some(value);
        self
    }

    pub(crate) fn server_address_value(&self) -> &str {
        &self.server_address
    }

    pub(crate) fn worker_target_value(&self) -> Option<&WorkerTarget> {
        self.worker_target.as_ref()
    }
}

impl Default for ClientOptions {
    fn default() -> Self {
        Self::new()
    }
}
